contrib/buildsystems/CMakeLists.txt

```diff
@@ -221,18 +221,22 @@ if(WIN32 AND NOT MSVC)#not required for visual studio builds
 	endif()
 endif()
 
-if(NO_GETTEXT)
-	message(STATUS "msgfmt not used under NO_GETTEXT")
-else()
+
+find_program(MSGFMT_EXE msgfmt)
+
+if(NOT MSGFMT_EXE)
+	if(WIN32)
+		include(NuGet)
+		nuget_install(Gettext.Tools)
```

Comment:
I wonder whether we can get a similar effect, albeit without having to use NuGet (which works very well for .NET projects but not well at all for C/C++ projects), if we used `vcpkg` for that? It [does have the `gettext` package](https://github.com/microsoft/vcpkg/tree/HEAD/ports/gettext), after all.

Reply:
NuGet is also used for build tools and utilities for various purposes.

I thought about using vcpkg for this, and you can build `gettext[tools]` to get `msgfmt`. However, gettext is a very long compile, and in this case we just want a utility not a link library, so I thought this would be a better way to go. I use this method in another project.

However, if you want to support one or the other or both I can do that.

Reply:
I'd rather not mix and match `vcpkg` and NuGet.

If you are really, really concerned, we can teach [this Azure Pipeline](https://dev.azure.com/git/git/_build?definitionId=9) to build `gettext[tools]` (or just `gettext`?) and offer it together with the other build artifacts.

Reply:
FWIW this Pipeline simply runs:

```powershell
& compat/vcbuild/vcpkg_install.bat
if (!$?) { exit(1); }

& compat/vcbuild/vcpkg_install.bat arm64-windows
if (!$?) { exit(1); }

git init vcbuild-artifacts\compat\vcbuild\vcpkg
if (!$?) { exit(1); }
Copy-Item compat\vcbuild\vcpkg\vcpkg.exe vcbuild-artifacts\compat\vcbuild\vcpkg -Force
if (!$?) { exit(1); }
Copy-Item compat\vcbuild\vcpkg\installed vcbuild-artifacts\compat\vcbuild\vcpkg -Force -Recurse
if (!$?) { exit(1); }
Copy-Item compat\vcbuild\vcpkg\packages vcbuild-artifacts\compat\vcbuild\vcpkg -Force -Recurse
if (!$?) { exit(1); }
# Copy-Item compat\vcbuild\vcpkg\scripts vcbuild-artifacts\compat\vcbuild\vcpkg -Force -Recurse
# if (!$?) { exit(1); }

# Remove manual pages (they are not used, and only bloat the artifact)
Get-ChildItem vcbuild-artifacts\compat\vcbuild\vcpkg html -recurse -directory | ForEach-Object {Remove-Item $_.FullName -Recurse -Force}
if (!$?) { exit(1); }
```

Therefore we would probably need to handle `NO_GETTEXT` in `vcpkg_install.bat`, too: if it is set, do what we do right now. If it is not set, include `gettext` in the build.

Reply:
My concern is making life less painful for contributors, when I say that gettext is a really long compile, I mean approximately half an hour on a typical system or more.

In my other vcpkg project, some of my contributors refused to use the VS support at all because of the long compile times.

But up to you.

This is a problem with vcpkg in general. Many people use Conan instead which has binary packages. I am planning to add Conan support to the other project I mentioned.

Speaking of compile times in a CI, vcpkg supports a new feature called binary caching: https://vcpkg.readthedocs.io/en/stable/users/binarycaching/

> we would probably need to handle NO_GETTEXT in vcpkg_install.bat

For the purposes of this PR, I can just export this as an env var. 

But this script needs some enhancements in general, like upgrading vcpkg itself and all dependencies.

Reply:
Among other enhancements, I was also thinking of having the vcpkg package list be passed in to the script on the commandline (not in this PR though.)

Reply:
Hmm. That caching sounded good, right until it talked about the `GITHUB_TOKEN`.

Maybe we should simply do our own caching. Like, by default (unless we're running in Azure Pipelines, that is), download the Azure Pipeline artifact instead of building `vcpkg`. Probably a new command-line option of `vcpkg_install.bat`... I just wonder how we would do this. You'd need the equivalent of `curl` and `unzip`. Either we expect Git to be installed, or we switch from a `.bat` to a `.ps1` script. I _think_ I'd like the latter more.

That would align regular users' scenarios with the way the CI works, and also would address the speed issues you mentioned.

Reply:
I have done some things with this, and it's possible to cache the `vcpkg/installed` directory. There are two disadvantages to this method:

- The tree is huge, extracting it takes quite a long time, depending on the number of packages installed. Not as long as building them however.
- This overwrites the whole vcpkg installation including the package database.

I was also working on this a couple of years ago:

https://github.com/rkitover/vcpkg-binpkg-prototype

, it did work more or less last time I was playing with it and I could fix it up etc.. I was going to set up a build job to make packages for my project, but I never got around to it.

This is all out of scope of this PR of course. I could replace the NuGet package with `gettext[tools]` if you like and pass in `NO_GETTEXT` as an env var.

> You'd need the equivalent of curl and unzip. 

Windows comes with `curl` and `tar` the past 2-3 years as standard utilities. The `tar` is a horrible `libarchive` implementation, but it does handle `.zip` files.

> switch from a .bat to a .ps1 script.

It can even be a `.cmake` script (callable from outside cmake using `cmake -P script.cmake` for non-cmake usage.) Whichever you prefer.

Reply:
I could also do the enhancement I mentioned here, that is pass the list of packages on the commandline, excluding `gettext[tools]` when necessary.

Some things mention the `Makefile` calling this script, but I could not find the invocation anywhere in the current sources.

Reply:
After mulling this over in the back of my mind, I think I have an even better idea: Instead of requiring `MSGFMT_EXE` unless `NO_GETTEXT` is set, how about implying the latter if the former isn't set/found? Something like:

```cmake
if(NOT NO_GETTEXT)
	find_program(MSGFMT_EXE msgfmt)

	if(NOT MSGFMT_EXE)
		message('No `msgfmt` found, assuming NO_GETTEXT')
		set(NO_GETTEXT TRUE)
	endif()
endif()
```

Reply:
Yeah it already works that way, I can also set that variable in the cache just in case.

Reply:
Actually looks like this is a bit more complex than that.

`NO_GETTEXT` is set if `libintl` is not found, but on `MSVC` it is always set because it does not use `libintl`, so I just broke this by making this change.

I will work on cleaning this up, let me know your thoughts on the other things I brought up and what you want me to do in this PR specifically regarding msgfmt.

Reply:
To clarify a bit further:

- `intl_FOUND` is set if `libintl` is available, but not on `MSVC` because it does not use libintl
- if `intl_FOUND` is not set, `-DNO_GETTEXT` is added, in which if `iconv_FOUND` isn't set either, than `-DNO_ICONV` is also added
- if `intl_FOUND` is not set, which is always the case for `MSVC`, then the **cmake** variable, and **not** the macro `NO_GETTEXT` is also set.
- The value of `NO_GETTEXT` is written to `GIT-BUILD-OPTIONS`, which will always be null for `MSVC`
- If `MSGFMT_EXE` is set, then translations are built, nothing else controls, the `NO_GETTEXT` variable has no effect on this.

I don't know what `GIT-BUILD-OPTIONS` is used for or whether locales even work for `MSVC` builds, I have only tried building the translations with `msgfmt` which does work.

Perhaps we should address this in a separate PR?

Reply:
I just verified that locales definitely **DO NOT** work for `MSVC` builds, even though translations are built now.

Reply:
@PhilipOakley @dscho 

I'm going to work on this possibly a few days more. It seems translations are not working on `MSVC` builds and there are a few other minor details to clean up.

Let me know your thoughts on the other points I brought up.

Reply:
> I don't know what GIT-BUILD-OPTIONS is used for

`GIT-BUILD-OPTIONS` is used e.g. in [the test suite](https://github.com/git-for-windows/git/blob/v2.35.1.windows.2/t/test-lib.sh#L66) (e.g. to skip the translation-related tests).

> I'm going to work on this possibly a few days more. It seems translations are not working on MSVC builds and there are a few other minor details to clean up.

Yes, let's focus on the common case where we simply won't need the translations.

> If MSGFMT_EXE is set, then translations are built, nothing else controls this, the NO_GETTEXT variable has no effect on this.

That looks like a bug to me. I would really like to avoid even looking for `MSGFMT_EXE` if `NO_GETTEXT` is set to `TRUE`, and if `NO_GETTEXT` is not set and `msgfmt` cannot be found, I want `NO_GETTEXT` to be set to `TRUE`.

In any case, I want the translations not to be built when `NO_GETTEXT` is set to `TRUE`.

Reply:
Could you please remove the need for NuGet, by respecting NO_GETTEXT to imply that MSGFMT_EXE is not needed?